src/gui/qgsmapcanvasannotationitem.cpp

```diff
@@ -120,6 +120,12 @@ void QgsMapCanvasAnnotationItem::updateBoundingRect()
 
 void QgsMapCanvasAnnotationItem::onCanvasLayersChanged()
 {
+  if ( !mMapCanvas->annotationsVisible() )
+  {
+    setVisible( false );
+    return;
```

Comment:
I'd add this as another branch in the checks rather than `return`ing here. Or return in the subsequent if as well.